### ISSUE TITLE
Proposal: Settings handling change

### ISSUE THREAD

**Issue by Sneddo** [alanrenouf/vCheck-vSphere#361]
So, I feel like we are hitting the limitations of the current settings and plugin structures. For example the ease (or lack thereof) of updates and the sheer number of settings to be configured, even if irrelevant.

What I've been considering is shifting the plugins from raw Powershell code, to an XML structure. This will greatly simplify the handling of settings, and remove the need to parse the text in the file.

e.g.

      <plugin>
         <info>
            <author>Author Name</author>
            <version>1.0</version>
            <category>vSphere</category>
            <title>Plugin title</title>
            <header>Plugin Header</header>
            <comments>Plugin Comments</comments>
            <display>Table</display>
         </info>
         <settings>
            <setting name='examplesetting' prompt='Do you want to set this setting?'>$true</setting>
            <setting name='settingdependancy' dependancy='$plugin.settings.examplesetting -eq $true' prompt='What value do you want to set?'>value</setting>
         </settings>
      
         <code>
            Write-Host "Plugincode goes here"
         </code>
      </plugin>

During execution, vCheck would read a settings.xml file (probably a modified version of the existing job XML), which would contain any overrides for default settings (so if defined it is used, otherwise use the default from the plugin). This would mean that we are no longer modifying the plugin file with plugin-specific settings and allow us to update the plugin code with no risk of losing existing configuration.

As you can see, this also allows us to define dependencies on settings- so for example if we were to extend this to the GlobalVariables.ps1 file we could make SMTP settings dependant on $SendEmail being true. We could also wrap in the EndScript.ps1 as a post-execution step.

The downside is that internationalization for the settings prompt is not straight-forward (though this is not currently supported anyway). Perhaps having a 'prompts' key, with culture-specific subkeys?

Due to the significant change, I'm proposing this as a 7.0 release, but keen to get some input on this.

**Comment by PaulWalkerUK** [comment]
So where each plugin is currently a "ps1" file, it would become an "xml" file with the actual Powershell code contained in the <code> element? It's certainly an interesting idea!

Off hand, the only thing that's got me thinking is whether there could be any problems embedding Powershell code within an XML element? I suppose it's unlikely the code would contain anything that looked like XML which might trip up the parser...

I'd been starting to think of something similar but using psd1 files to contain prompts in different languages (as per standard Powershell functionality) and probably another psd1 file to contain user-chosen values. I think I prefer your suggestion though - I especially like the idea of declaring dependencies. 

I agree about it being a new major version.

**Comment by alanrenouf** [comment]
Personally I prefer the PS1 rather than XML, I like being able to run them independently and quicker and easier debug, i think introducing an XML format would make it harder for people to pick up and contribute back to, something that makes this popular today!

**Comment by monahancj** [comment]
I agree with Alan about both sides of that coin.  I think embedding in
XML would create a barrier to entry for people new to PowerShell.  It
wouldn't stop me, but I have little enough extra time to contribute to that
adding complexity means I would get less done in the time I do have.

  I think putting the variables in a separate file would be helpful.
Separating the data from the code.  I've been looking at how to easily
upgrade to a new version of vCheck without having to manually reenter all
the settings, and how to easily maintain similar settings across multiple
vCheck report instances.  We have 3 large VMare environments and several
small remote environments each with their own vCheck report.

  If the settings were broken out into a powershell manageable format like
XML or CSV then a few functions could be written to manage the settings.
Like Get-VcheckSettings -Path blahblah -PluginName blahblah |
Set-VcheckSettings -Path diffpath -Plugin blahblah.  Having the functions
written to manage the more complex machine readable settings files would
hide the added complexity and keep the barrier to entry low.

**Comment by Sneddo** [comment]
Yep, valid point. Exactly why I started this issue to get some discussion happening :)

I don't think introducing a separate settings file per plugin is the answer- I like that plugins are effectively self-contained. Though having a "Get-vCheckSetting" type of function could work...if we call it after the default variables are set and override if a different value is found in a XML/CSV data structure. 

Perhaps we could extend the existing comment/variable combination to incorporate dependencies, variable types, validation rules, etc.

**Comment by monahancj** [comment]
That could work.  It leaves people the option of continuing the old way
or converting at their convenience to the new way.

1- Set variables defined in the plugin.ps1
2- Set, possibly overriding, variables defined in the xml/csv file.

  Having a single xml/csv file for all the plugin variables would be easier
to manage for upgrades and new installs.  Copy one file instead of dozens.

  If the variables file was CSV it would be easy to write the get/set
functions, and it would be easy for users to open up a text file or
spreadsheet software to make changes.  If xml was used I think the get/set
functions would need to be released at the same time.  The only people I
know that are comfortable working with xml are devs and sysadmins with a
dev bent.  As of now complete novices that move beyond using the wizard can
easily manage vCheck with Notepad.  I think we should try to stick as close
to that as we can.

**Comment by Sneddo** [comment]
Yeah, the benefit of XML is that can be directly queried, but in this case CSV could be better for ease of use. The benefit of using XML for the settings was to structure the data with extensible properties for validation/prerequisites/etc, but if we aren't doing that then it becomes irrelevant.

I guess the CSV would need plugin name, setting name and value. 

So I think there are two parts to this:
1. Create the Get-vCheckSetting/Set-vCheckSetting functions - this solves the immediate problem of updates. Gradually roll this out to all plugins.
2. Extend setting handling to include nice to have functionality (prerequisites, validation) - this solves the current limitations of the settings handling.



We could actually merge the settings from the global variables into vCheck.ps1 as well, since we will effectively replace GlobalVariables.ps1 with settings.csv (or whatever we end up naming it). For a 7.0 release we could be nice and include a check for GlobalVariables.ps1 and do the conversion and delete the ps1. Would mean missing out on plugin settings, but this would be too difficult IMO as it would need to be checked against an unmodified version.

**Comment by vScripter** [comment]
I also agree that it's a catch-22. If it were me, I like the extensibility that XML brings to the table, which could make it easier for modularity/furture integrations. That being said, for the sake of the majority of people that use this tool, reviewing XML can certainly be anti-pattern for readability/ease of understanding.

I really like the idea of baking in the Get/Check-vCheckSettings functions. (I've actually been longing for a change to work on making that process a bit more transparent).

To further the enhancements for a 7.0 release, what about including an optional 'Settings' GUI that could leverage underlying functions to review/set options? Nothing too fancy or difficult, maybe just a basic Windows Form; again, the point being that it would be optional and could help bridge the gap and get people up and running a bit faster. If it hasn't been thought of, already (which would be a surprise), or it was started but never finished, I have free cycles to help kick the tires on such a movement. 

Either way, I agree/support the overall initiative of refactoring the 'settings framework'.

**Comment by Sneddo** [comment]
Yep, I was working on a proof of concept config GUI that I posted about in Issue #270.

I've since changed jobs and not had much spare time to work on it, and then I just forgot about it :(
https://github.com/Sneddo/vCheck-Tools

If you have spare time, it would be much appreciated. Probably worth creating a feature branch for it and merging into the vCheck repo (will actually do that now)...feel free to use or scrap any of it if you have time to work on it, as I said it was more of a proof of concept than anything.
